[NativeAOT] allow creating CCW for resurrected objects

AustinWise opened the pull request:
This PR adds a test that on main passes for CoreCLR but segfaults for NativeAOT. It also includes fixes to make the test pass for NativeAOT.

The key problem is the conditional weak table will keep the `ManagedObjectWrapperHolder` alive but not prevent it from being finalized. This means that if the wrapped object is resurrected, calls to `GetOrCreateComInterfaceForObject` could return a pointer freed memory.

The solution this PR implements is to move the finialization of the `ManagedObjectWrapper` to a separate class. This class defers finialization until the wrapped object has been completely collected. The new `ManagedObjectWrapperReleaser` is referenced by the `ManagedObjectWrapperHolder` class, so it should not be finialized until the wrapped object becomes eligible for finialization.

The references between objects looks like this:

```mermaid
graph TD
WO[wrapped object];
MOW["ManagedObjectWrapper (unmanaged struct)"];
MOWH[ManagedObjectWrapperHolder];
Releaser[ManagedObjectWrapperReleaser];
WO -->|conditional weak table| MOWH;
MOWH --> WO;
MOWH --> MOW;
MOW -->|ref-counted handle| MOWH;
Releaser --> MOW;
MOWH --> Releaser;
```

Some other race conditions are also fixed by this PR:
* Previously the ref-counted handle in the `ManagedObjectWrapper` was not set until after the `ManagedObjectWrapperHolder` was added to the CWT. Hypothetically if a holder was added to the CWT but not yet initialized with the ref-counted handle, another thread could read the object out of the CWT and observe the null handle. This is fixed by doing all the initialization of the holder in its constructor.
* The `GetOrCreateComInterfaceForObject` was assuming that `ConditionalWeakTable.GetValue` always returns a new object and calling `AddRef` on the holder was not needed. However, `GetValue` can potentially call `TryGetValue` and return an existing instance. This means that if many threads called `GetOrCreateComInterfaceForObject` at the same time, the ref count on the `ManagedObjectWrapper` might not be as high as it should be. This is fixed by unconditionally calling `AddRef` on the holder.

AustinWise committed:
[NativeAOT] allow creating CCW for resurrected objects

AustinWise commented:
FYI: this is still a draft because while this PR does make the tests pass, it introduces race conditions. Specifically there is no coordination between a thread resurrecting the holder & calling `AddRef` with the finalizer thread. So it is still possible for the strong reference to be released on the wrapped object while the COM reference count is non-zero.

AustinWise committed:
Split the responsbility of keeping the objects alive

The WrappedObjectHolder will be used to keep the managed object alive when
the COM reference count is positive. And the ManagedObjectWrapper object
will be responsible for cleaning up the unmanaged MOW when the wrapped
object is truly dead.

The design is that as long as there is a postive reference count, the
WrappedObjectHolder should stay alive. Any time the ref count is increased
from the managed side from 0 to 1, the WrappedObjectHolder will be restored
if it was collected. It should not be possible for the MOW to observe a
null WrappedObjectHolder.

AustinWise commented:
Instead of having one object for managing lifetimes of both the managed and unmanaged objects, I split the responsibilities into two objects. I think this is closer to a solution, but I have to review it a bit more and write down my assumptions about how GCHandle work.

AustinWise committed:
Use a DependentHandle to keep the WrappedObjectHolder alive.

I'm still not 100% sure that _exposed is thread safe.

AustinWise commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs:
```diff
@@ -404,42 +404,62 @@ static bool IsRootedCallback(IntPtr pObj)
-            public uint AddRef() => _wrapper->AddRef();
+            public uint AddRef(object wrappedObject)
+            {
+                _exposed = true;
```
I'm not sure if I need more synchronization on this variable. It is set to true on any random thread and read from the finializer thread.

AustinWise replied:
I'm just going to remove this for now. I'll revisit the issue of extra memory being used when there is a race condition inserting into the conditional weak table later.

AaronRobinsonMSFT commented on src/tests/Interop/COM/ComWrappers/API/Program.cs:
```diff
@@ -336,6 +336,62 @@ unsafe static void CallSetValue(TestComWrappers wrappers, Test testInstance, int
+                IntPtr itestPtr;
+                Assert.Equal(0, Marshal.QueryInterface(nativeInstance, ref iid, out itestPtr));
+
+                var inst = Marshal.PtrToStructure<VtblPtr>(itestPtr);
```
We shouldn't be using `Marshal.PtrToStructure()` anywhere in here. Please defer to pointers and remove the other two uses as well.

AustinWise replied:
I switch to using pointers.

AustinWise commented:
> Based on a quick look at the test I'm not sure I understand what this is validating.

On the main branch, the test passes on CoreClr but segfaults on NativeAOT. This is because the `ManagedObjectWrapperHolder` gets finalized and frees the `ManagedObjectWrapper` too early. I have updated the description of the PR to be a little more clear about this.

AaronRobinsonMSFT commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs:
```diff
@@ -404,42 +404,62 @@ static bool IsRootedCallback(IntPtr pObj)
+                if (_exposed && _wrappedObject.Target != null)
+                {
+                    // The wrapped object has not been fully collected, so it is still
+                    // potentially reachable via the CWT. Keep ourselves alive in case
```
What is the CWT acronym?

jkoritzinsky replied:
ConditionalWeakTable

AaronRobinsonMSFT replied:
I see. There is no need to use an acronym here.

AustinWise replied:
I fixed the comment.

AustinWise committed:
Respond to feedback

AustinWise committed:
Simplify object ownership structure.

AustinWise committed:
Restore the call to _ccwTable.TryGetValue.

This is to avoid allocating a closure if possible.

AustinWise commented on src/tests/Interop/COM/ComWrappers/API/Program.cs:
```diff
@@ -336,6 +336,62 @@ unsafe static void CallSetValue(TestComWrappers wrappers, Test testInstance, int
+                // release for QueryInterface
+                Assert.Equal(1, Marshal.Release(itestPtr));
+                // release for GetOrCreateComInterfaceForObject
+                Assert.Equal(0, Marshal.Release(itestPtr));
```
Thanks, I fixed it.

AustinWise committed:
Also apply the feedback to the correct unit test.

This code was copy-pasted.

AustinWise commented:
I updated how the responsibility is split for keeping objects alive and collecting them. I think this new design is much easier to understand. The description of the PR has been updated.

The main problem this PR does not address is the issue of extra ManagedObjectWrappers hanging around for the lifetime of a wrapped object when there is a race condition adding values to the `_ccwTable`.

AustinWise committed:
readonly

AustinWise committed:
Use refcounted handle to asses the lifetime of the holder

This saves a handle. The refcounted handle behaves like a long-weak,
so this should not get cleared out until the conditional weak table
is also cleared out.

AustinWise commented:
I realized that I did not need to use an extra handle to monitor the lifetime of the holder object from the releaser object. Instead I could just use the ref counted handle. That saves allocating an extra GC handle.

jkoritzinsky reviewed (approved):
This looks good to me, but I want @AaronRobinsonMSFT to take another look.

AaronRobinsonMSFT commented:
> The key problem is the conditional weak table will keep the ManagedObjectWrapperHolder alive but not prevent it from being finalized. 

My biggest issue here is the above statement. If an object is "alive", I assume "reachable", then it will not be finalized. The GC only adds objects to the finalizer queue if they are no longer reachable so I am confused by this statement.

AaronRobinsonMSFT commented:
> > The key problem is the conditional weak table will keep the ManagedObjectWrapperHolder alive but not prevent it from being finalized.
> 
> My biggest issue here is the above statement. If an object is "alive", I assume "reachable", then it will not be finalized. The GC only adds objects to the finalizer queue if they are no longer reachable so I am confused by this statement.

Oh right. The conditional weak table doesn't keep it alive, merely a reference to acquire. Its state is indeterminant in this case and that is the issue you are attempting to mitigate. This is simply loose terminology.
